// mongo-native/lib/collection-interface.js
var helpers = require('./helpers');

var CollectionInterface = helpers.createClass({
	collectionName: helpers.notImplemented('collectionName'),
  namespace: helpers.notImplemented('namespace'),
  writeConcern: helpers.notImplemented('writeConcern'),
  hint: helpers.notImplemented('hint'),
  find: helpers.notImplemented('find'),
  insertOne: helpers.notImplemented('insertOne'),
  insertMany: helpers.notImplemented('insertMany'),
  bulkWrite: helpers.notImplemented('bulkWrite'),
  insert: helpers.notImplemented('insert'),
  updateOne: helpers.notImplemented('updateOne'),
  replaceOne: helpers.notImplemented('replaceOne'),
  updateMany: helpers.notImplemented('updateMany'),
  update: helpers.notImplemented('update'),
  deleteOne: helpers.notImplemented('deleteOne'),
  removeOne: helpers.notImplemented('removeOne'),
  deleteMany: helpers.notImplemented('deleteMany'),
  removeMany: helpers.notImplemented('removeMany'),
  remove: helpers.notImplemented('remove'),
  save: helpers.notImplemented('save'),
  findOne: helpers.notImplemented('findOne'),
  rename: helpers.notImplemented('rename'),
  drop: helpers.notImplemented('drop'),
  options: helpers.notImplemented('options'),
  isCapped: helpers.notImplemented('isCapped'),
  createIndex: helpers.notImplemented('createIndex'),
  createIndexes: helpers.notImplemented('createIndexes'),
  dropIndex: helpers.notImplemented('dropIndex'),
  dropIndexes: helpers.notImplemented('dropIndexes'),
  dropAllIndexes: helpers.notImplemented('dropAllIndexes'),
  reIndex: helpers.notImplemented('reIndex'),
  listIndexes: helpers.notImplemented('listIndexes'),
  ensureIndex: helpers.notImplemented('ensureIndex'),
  indexExists: helpers.notImplemented('indexExists'),
  indexInformation: helpers.notImplemented('indexInformation'),
  count: helpers.notImplemented('count'),
  distinct: helpers.notImplemented('distinct'),
  indexes: helpers.notImplemented('indexes'),
  stats: helpers.notImplemented('stats'),
  findOneAndDelete: helpers.notImplemented('findOneAndDelete'),
  findOneAndReplace: helpers.notImplemented('findOneAndReplace'),
  findOneAndUpdate: helpers.notImplemented('findOneAndUpdate'),
  findAndModify: helpers.notImplemented('findAndModify'),
  findAndRemove: helpers.notImplemented('findAndRemove'),
  aggregate: helpers.notImplemented('aggregate'),
  parallelCollectionScan: helpers.notImplemented('parallelCollectionScan'),
  geoNear: helpers.notImplemented('geoNear'),
  geoHaystackSearch: helpers.notImplemented('geoHaystackSearch'),
  group: helpers.notImplemented('group'),
  mapReduce: helpers.notImplemented('mapReduce'),
  initializeUnorderedBulkOp: helpers.notImplemented('initializeUnorderedBulkOp'),
  initializeOrderedBulkOp: helpers.notImplemented('initializeOrderedBulkOp')
});

module.exports = CollectionInterface;
